[UnifiedPDF] Unify the "partial" and "full" PDF tile code paths

smfr opened the pull request:
#### 74f0ad1b6f48586b34a55482948f69c7596d0a34
<pre>
[UnifiedPDF] Unify the &quot;partial&quot; and &quot;full&quot; PDF tile code paths
<a href="https://bugs.webkit.org/show_bug.cgi?id=272945">https://bugs.webkit.org/show_bug.cgi?id=272945</a>
<a href="https://rdar.apple.com/126723009">rdar://126723009</a>

Reviewed by Tim Horton.

AsyncPDFRenderer had separate code paths for partial and full tile updates, with two places
where we called `m_paintingWorkQueue-&gt;dispatch(...)`. Remove the one in `AsyncPDFRenderer::pdfContentChangedInRect()`,
calling `enqueueTilePaintIfNecessary()` instead.

`AsyncPDFRenderer::enqueueTilePaintIfNecessary()` now takes an optional clipRect for partial updates.
We need to take care when there is already a paint for this tile in flight; there&apos;s no guarantee that
it will finish before our new one, so we have just override it, taking care to unite any partial updates.
The newly added PDFTileRenderIdentifier allows us to determine, when a tile render completes, if it&apos;s the
most recent one.

Now that we have this tracking, we can also remove entries from `m_currentValidTileRenders` as soon
as they are complete, rather than waiting until paint time, which was a poor version of batching.

Also ensure that `paintTileOnWorkQueue()` always notifies the main thread, even in the case of buffer
allocation failure, to avoid leaving entries in m_currentValidTileRenders.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
(WebKit::AsyncPDFRenderer::enqueueTilePaintIfNecessary):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::willRepaintTile):
(WebKit::AsyncPDFRenderer::willRemoveTile):
(WebKit::AsyncPDFRenderer::willRepaintAllTiles): Just call clearRequestsAndCachedTiles().
(WebKit::AsyncPDFRenderer::clearRequestsAndCachedTiles):
(WebKit::AsyncPDFRenderer::enqueueTilePaintIfNecessary):
(WebKit::AsyncPDFRenderer::renderInfoForTile const):
(WebKit::AsyncPDFRenderer::enqueuePaintWithClip):
(WebKit::AsyncPDFRenderer::paintTileOnWorkQueue):
(WebKit::AsyncPDFRenderer::transferBufferToMainThread):
(WebKit::AsyncPDFRenderer::didCompleteTileRender):
(WebKit::AsyncPDFRenderer::paintTilesForPage):
(WebKit::AsyncPDFRenderer::pdfContentChangedInRect):
(WebKit::AsyncPDFRenderer::didCompleteNewTileRender): Deleted.
(WebKit::AsyncPDFRenderer::didCompleteTileUpdateRender): Deleted.

Canonical link: <a href="https://commits.webkit.org/277747@main">https://commits.webkit.org/277747@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2870ae571b971115b04c3a174977d582722a8c67

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/48379 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/27591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/51337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/51067 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/44444 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/33527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/25112 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/39535 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48961 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/25277 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41770 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20665 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22757 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/6435 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44706 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/43415 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52972 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/23426 "Built successfully") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/19758 "Build is in progress. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; 14 flakes 3 failures; Uploaded test results; 13 flakes 4 failures; Running compile-webkit-without-change") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/46856 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24691 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/41964 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/45771 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10685 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/25496 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/24414 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fbbba21b5172549ee796fbb4db2189be8022b4ff. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/27491

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2870ae571b971115b04c3a174977d582722a8c67. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/27491

smfr committed:
[UnifiedPDF] Unify the "partial" and "full" PDF tile code paths
https://bugs.webkit.org/show_bug.cgi?id=272945
rdar://126723009

Reviewed by Tim Horton.

AsyncPDFRenderer had separate code paths for partial and full tile updates, with two places
where we called `m_paintingWorkQueue->dispatch(...)`. Remove the one in `AsyncPDFRenderer::pdfContentChangedInRect()`,
calling `enqueueTilePaintIfNecessary()` instead.

`AsyncPDFRenderer::enqueueTilePaintIfNecessary()` now takes an optional clipRect for partial updates.
We need to take care when there is already a paint for this tile in flight; there's no guarantee that
it will finish before our new one, so we have just override it, taking care to unite any partial updates.
The newly added PDFTileRenderIdentifier allows us to determine, when a tile render completes, if it's the
most recent one.

Now that we have this tracking, we can also remove entries from `m_currentValidTileRenders` as soon
as they are complete, rather than waiting until paint time, which was a poor version of batching.

Also ensure that `paintTileOnWorkQueue()` always notifies the main thread, even in the case of buffer
allocation failure, to avoid leaving entries in m_currentValidTileRenders.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.h:
(WebKit::AsyncPDFRenderer::enqueueTilePaintIfNecessary):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/AsyncPDFRenderer.mm:
(WebKit::AsyncPDFRenderer::willRepaintTile):
(WebKit::AsyncPDFRenderer::willRemoveTile):
(WebKit::AsyncPDFRenderer::willRepaintAllTiles): Just call clearRequestsAndCachedTiles().
(WebKit::AsyncPDFRenderer::clearRequestsAndCachedTiles):
(WebKit::AsyncPDFRenderer::enqueueTilePaintIfNecessary):
(WebKit::AsyncPDFRenderer::renderInfoForTile const):
(WebKit::AsyncPDFRenderer::enqueuePaintWithClip):
(WebKit::AsyncPDFRenderer::paintTileOnWorkQueue):
(WebKit::AsyncPDFRenderer::transferBufferToMainThread):
(WebKit::AsyncPDFRenderer::didCompleteTileRender):
(WebKit::AsyncPDFRenderer::paintTilesForPage):
(WebKit::AsyncPDFRenderer::pdfContentChangedInRect):
(WebKit::AsyncPDFRenderer::didCompleteNewTileRender): Deleted.
(WebKit::AsyncPDFRenderer::didCompleteTileUpdateRender): Deleted.

Canonical link: https://commits.webkit.org/277747@main

webkit-commit-queue commented:
Committed 277747@main (74f0ad1b6f48): <https://commits.webkit.org/277747@main>

Reviewed commits have been landed. Closing PR #27491 and removing active labels.
